### PR TITLE
Add MultiNamespaceListerWatcher capability

### PR DESCRIPTION
**Pull request by clyang82**
Add MultiNamespaceListerWatcher method into top `pkg` which can be used by
#11667 
#11932 

Signed-off-by: Chun Lin Yang <clyang@cn.ibm.com>

### CODE REVIEW

**Comment by clyang82** [comment]
have a separate commit for new dependency instead of create a new PR. so close it.

File: pkg/listwatch/listwatch.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+// ResultChan implements the watch.Interface interface.
+func (mw *multiWatch) ResultChan() <-chan watch.Event {
+	return mw.result
+}
+
+// Stop implements the watch.Interface interface.
+// It stops all of the underlying watch.Interfaces and closes the backing chan.
+// Can safely be called more than once.
+func (mw *multiWatch) Stop() {
+	select {
```
Comment (ozevren) [inline pkg/listwatch/listwatch.go]: There is a race condition here, which can happen when Stop gets called reentrantly. stop methods can be called multiple times. Not sure if this is a problem with the sub-watchers.

File: pkg/listwatch/listwatch.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
+// Copyright 2019 Istio Authors
```
Comment (ozevren) [inline pkg/listwatch/listwatch.go]: A couple of notes:
I am not sure that this is the right abstraction to create. The fact that you had to come up with your own demultiplexing scheme for the ListOptions is an indicator of that. IMHO coming up with an infrastructure that deals with the life-cycle aspects of ListerWatchers that talk to a common SharedInformerFactory is a better approach.

Also, I don't think we want this in top level pkg. Moving forward, Galley will be the only component that will talk with Kubernetes. We should move this to galley/pkg/source/kube/common.

I've some review comments as well. Hope this helps.

File: pkg/listwatch/listwatch.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+)
+
+// MultiNamespaceListerWatcher takes a list of namespaces and a
+// cache.ListerWatcher generator func and returns a single cache.ListerWatcher
+// capable of operating on multiple namespaces.
+func MultiNamespaceListerWatcher(namespaces []string, f func(string) cache.ListerWatcher) cache.ListerWatcher {
```
Comment (ozevren) [inline pkg/listwatch/listwatch.go]: Nit: f => fn.
Also consider creating a type for it (i.e. CreateListerWatcherFn)

File: pkg/listwatch/listwatch.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+	return &l, nil
+}
+
+// Watch implements the ListerWatcher interface.
+// It returns a watch.Interface that combines the output from the
+// watch.Interface of every cache.ListerWatcher into a single result chan.
+func (mlw multiListerWatcher) Watch(options metav1.ListOptions) (watch.Interface, error) {
+	resourceVersions := make([]string, len(mlw))
+	// Allow resource versions to be "".
+	if options.ResourceVersion != "" {
```
Comment (ozevren) [inline pkg/listwatch/listwatch.go]: This is extremely counter intuitive and breaks the regular contract of options. Not supporting a starting resource version (i.e. erroring out) is a simpler choice here.

File: pkg/listwatch/listwatch.go (hunk 2 above)
Comment (clyang82) [reply pkg/listwatch/listwatch.go]: @ozevren Thanks for your time to review this PR. What is the status for Galley will be the only component that will talk with k8s? will it be in 1.2?

**Comment by ozevren** [comment]
Realistically speaking 1.3.

**Comment by clyang82** [comment]
> Realistically speaking 1.3.

I should move this to galley/pkg/source/kube/common and address your comment. then it can be in 1.2. Right?

**Comment by duderino** [comment]
Judging by the code review comments, too much work would be required for 1.1.  Please try resending this to master and addressing the review comments there.

**Pull request closed without merging**